executor/src/witgen/evaluators/expression_evaluator.rs

```diff
@@ -0,0 +1,133 @@
+use std::collections::BTreeMap;
+
+use powdr_ast::analyzed::{
+    AlgebraicBinaryOperation, AlgebraicBinaryOperator, AlgebraicExpression as Expression,
+    AlgebraicReference, AlgebraicReferenceThin, AlgebraicUnaryOperation, AlgebraicUnaryOperator,
+    Analyzed, PolyID, PolynomialType,
+};
+use powdr_number::FieldElement;
+
+/// Accessor for trace values.
+pub trait TraceValues<T> {
+    fn get(&self, poly_id: &AlgebraicReference) -> T;
+}
+
+/// A simple container for trace values.
+pub struct OwnedTraceValues<T> {
+    pub values: BTreeMap<PolyID, Vec<T>>,
```

Comment:
Two notes:
- Making this owned is not ideal, because it typically requires copying the trace. But I think this happens already before this PR, and part of the reason is that we currently add a dummy column in `backend_utils::machine_witness_columns`. I think in the future we can only keep references here.
- We should consider moving to a `ColumnMap` (which is essentially a `Vec`) instead of a `BTreeMap` here. I don't want to do it here though, because then we'd have to require having values for all columns.